library/core/src/ptr/mod.rs

```diff
@@ -883,12 +883,18 @@ pub const unsafe fn read_unaligned<T>(src: *const T) -> T {
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub unsafe fn write<T>(dst: *mut T, src: T) {
-    if cfg!(debug_assertions) && !is_aligned_and_not_null(dst) {
-        // Not panicking to keep codegen impact smaller.
-        abort();
+    // We are calling the intrinsics directly to avoid function calls in the generated code.
+    extern "rust-intrinsic" {
+        fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize);
+    }
+
+    // SAFETY: the caller must guarantee that `dst` is valid for writes.
+    // `dst` cannot overlap `src` because the caller has mutable access
+    // to `dst` while `src` is owned by this function.
+    unsafe {
+        copy_nonoverlapping(&src as *const T, dst, 1);
+        intrinsics::forget(src);
```

Comment:
What is the difference between `intrinsics::forget` and `mem::forget`? I saw that `mem::forget` is used by `write_unaligned` and `mem::forget` is `const fn`

Reply:
`mem::forget` uses `intrinsics::forget` internally, but doesn't get inlined in debug mode.

Reply:
Ah, ok. Thanks

Reply:
Yeah, currently only `write` gets this special treatment.

We could also decide to make `copy_nonoverlapping` directly an intrinsic again without a wrapper function -- the debug assertions in there are disabled now anyway, and that would benefit all the functions in here, not just `write`.

Reply:
Am I right in assuming that we will want to make `write` `const` in a not too distant future? In that case is there anything in this PR to consider to prevent any future issues?

I believe borrowing `src` might be a problem in const context until insert_name_here is implemented. Do we think that is relevant to this PR?

Reply:
> I believe borrowing src might be a problem in const context until insert_name_here is implemented

On second thought, is that #80418? :)

Reply:
While yes, we want to make it `const`, I also don't think we should make our runtime implementation worse, or keep such a complex intrinsic around, just to be able to make it `const`.

But also, as you said, with https://github.com/rust-lang/rust/pull/80418 this function should actually be fine.

Reply:
> While yes, we want to make it const, **I also don't think we should make our runtime implementation worse, or keep such a complex intrinsic around, just to be able to make it const**.

Not what what I meant to suggest :)

> But also, as you said, with #80418 this function should actually be fine.

Awesome then :)